File: mobile/lib/models/relatorio_alunos_turma.dart
class RelatorioAlunosTurmaAluno {
  const RelatorioAlunosTurmaAluno({
    required this.inscricaoId,
    required this.alunoNome,
    this.alunoCpf,
    this.alunoCpfFormatado,
    this.alunoIdade,
    this.escolaridadeDescricao,
    required this.dataInscricao,
    this.dataMatricula,
    this.dataCancelamento,
    required this.situacao,
    required this.situacaoCodigo,
    required this.dataUltSituacao,
    required this.quantidadeAtendimentos,
  });

  final String inscricaoId;
  final String alunoNome;
  final String? alunoCpf;
  final String? alunoCpfFormatado;
  final int? alunoIdade;
  final String? escolaridadeDescricao;
  final String dataInscricao;
  final String? dataMatricula;
  final String? dataCancelamento;
  final String situacao;
  final String situacaoCodigo;
  final String dataUltSituacao;
  final int quantidadeAtendimentos;

  factory RelatorioAlunosTurmaAluno.fromJson(Map<String, dynamic> json) {
    int? parseIdade(dynamic v) {
      if (v == null) return null;
      if (v is int) return v;
      return int.tryParse(v.toString());
    }

    return RelatorioAlunosTurmaAluno(
      inscricaoId: json['inscricaoId'] as String,
      alunoNome: json['alunoNome'] as String,
      alunoCpf: json['alunoCpf'] as String?,
      alunoCpfFormatado: json['alunoCpfFormatado'] as String?,
      alunoIdade: parseIdade(json['alunoIdade']),
      escolaridadeDescricao: json['escolaridadeDescricao'] as String?,
      dataInscricao: json['dataInscricao'] as String,
      dataMatricula: json['dataMatricula'] as String?,
      dataCancelamento: json['dataCancelamento'] as String?,
      situacao: json['situacao'] as String,
      situacaoCodigo: json['situacaoCodigo'] as String,
      dataUltSituacao: json['dataUltSituacao'] as String,
      quantidadeAtendimentos: json['quantidadeAtendimentos'] as int? ?? 0,
    );
  }
}

class RelatorioAlunosTurmaResumoSecao {
  const RelatorioAlunosTurmaResumoSecao({
    required this.matriculados,
    required this.matriculasCanceladas,
    this.aMatricular,
  });

  final int matriculados;
  final int matriculasCanceladas;
  final int? aMatricular;

  factory RelatorioAlunosTurmaResumoSecao.fromJson(Map<String, dynamic> json) {
    return RelatorioAlunosTurmaResumoSecao(
      matriculados: json['matriculados'] as int? ?? 0,
      matriculasCanceladas: json['matriculasCanceladas'] as int? ?? 0,
      aMatricular: json['aMatricular'] as int?,
    );
  }
}

class RelatorioAlunosTurmaSecao {
  const RelatorioAlunosTurmaSecao({
    required this.cursoCodigo,
    required this.cursoDescricao,
    required this.turmaCodigo,
    required this.turmaNome,
    required this.periodo,
    this.periodoRotulo,
    required this.situacaoTurma,
    this.situacaoTurmaRotulo,
    required this.alunos,
    required this.resumo,
  });

  final int cursoCodigo;
  final String cursoDescricao;
  final int turmaCodigo;
  final String turmaNome;
  final String periodo;
  final String? periodoRotulo;
  final String situacaoTurma;
  final String? situacaoTurmaRotulo;
  final List<RelatorioAlunosTurmaAluno> alunos;
  final RelatorioAlunosTurmaResumoSecao resumo;

  factory RelatorioAlunosTurmaSecao.fromJson(Map<String, dynamic> json) {
    int parseCodigo(dynamic v) {
      if (v is int) return v;
      return int.parse(v.toString());
    }

    return RelatorioAlunosTurmaSecao(
      cursoCodigo: parseCodigo(json['cursoCodigo']),
      cursoDescricao: json['cursoDescricao'] as String,
      turmaCodigo: parseCodigo(json['turmaCodigo']),
      turmaNome: json['turmaNome'] as String,
      periodo: json['periodo'] as String,
      periodoRotulo: json['periodoRotulo'] as String?,
      situacaoTurma: json['situacaoTurma'] as String,
      situacaoTurmaRotulo: json['situacaoTurmaRotulo'] as String?,
      alunos: (json['alunos'] as List<dynamic>? ?? [])
          .map(
            (e) => RelatorioAlunosTurmaAluno.fromJson(
              e as Map<String, dynamic>,
            ),
          )
          .toList(),
      resumo: RelatorioAlunosTurmaResumoSecao.fromJson(
        json['resumo'] as Map<String, dynamic>,
      ),
    );
  }
}

class RelatorioAlunosTurmaResumoGeral {
  const RelatorioAlunosTurmaResumoGeral({
    required this.cursoCodigo,
    required this.cursoDescricao,
    required this.turmaCodigo,
    required this.turmaNome,
    this.periodoRotulo,
    required this.situacaoTurma,
    this.situacaoTurmaRotulo,
    required this.matriculados,
    required this.matriculasCanceladas,
    this.aMatricular,
  });

  final int cursoCodigo;
  final String cursoDescricao;
  final int turmaCodigo;
  final String turmaNome;
  final String? periodoRotulo;
  final String situacaoTurma;
  final String? situacaoTurmaRotulo;
  final int matriculados;
  final int matriculasCanceladas;
  final int? aMatricular;

  factory RelatorioAlunosTurmaResumoGeral.fromJson(Map<String, dynamic> json) {
    int parseCodigo(dynamic v) {
      if (v is int) return v;
      return int.parse(v.toString());
    }

    return RelatorioAlunosTurmaResumoGeral(
      cursoCodigo: parseCodigo(json['cursoCodigo']),
      cursoDescricao: json['cursoDescricao'] as String,
      turmaCodigo: parseCodigo(json['turmaCodigo']),
      turmaNome: json['turmaNome'] as String,
      periodoRotulo: json['periodoRotulo'] as String?,
      situacaoTurma: json['situacaoTurma'] as String,
      situacaoTurmaRotulo: json['situacaoTurmaRotulo'] as String?,
      matriculados: json['matriculados'] as int? ?? 0,
      matriculasCanceladas: json['matriculasCanceladas'] as int? ?? 0,
      aMatricular: json['aMatricular'] as int?,
    );
  }
}

class RelatorioAlunosTurmaResult {
  const RelatorioAlunosTurmaResult({
    required this.todosCursos,
    required this.todasTurmas,
    this.cursoCodigo,
    this.turmaCodigo,
    required this.situacao,
    required this.secoes,
    required this.resumoGeral,
  });

  final bool todosCursos;
  final bool todasTurmas;
  final int? cursoCodigo;
  final int? turmaCodigo;
  final String situacao;
  final List<RelatorioAlunosTurmaSecao> secoes;
  final List<RelatorioAlunosTurmaResumoGeral> resumoGeral;

  factory RelatorioAlunosTurmaResult.fromJson(Map<String, dynamic> json) {
    final filtros = json['filtros'] as Map<String, dynamic>;

    int? parseCodigoOpt(dynamic v) {
      if (v == null) return null;
      if (v is int) return v;
      return int.tryParse(v.toString());
    }

    return RelatorioAlunosTurmaResult(
      todosCursos: filtros['todosCursos'] as bool? ?? false,
      todasTurmas: filtros['todasTurmas'] as bool? ?? false,
      cursoCodigo: parseCodigoOpt(filtros['cursoCodigo']),
      turmaCodigo: parseCodigoOpt(filtros['turmaCodigo']),
      situacao: filtros['situacao'] as String? ?? 'TODAS',
      secoes: (json['secoes'] as List<dynamic>? ?? [])
          .map(
            (e) => RelatorioAlunosTurmaSecao.fromJson(
              e as Map<String, dynamic>,
            ),
          )
          .toList(),
      resumoGeral: (json['resumoGeral'] as List<dynamic>? ?? [])
          .map(
            (e) => RelatorioAlunosTurmaResumoGeral.fromJson(
              e as Map<String, dynamic>,
            ),
          )
          .toList(),
    );
  }
}

File: backend/src/services/relatorio-alunos-turma.ts
import type { Prisma } from "@prisma/client";

import { formatDataBr } from "../lib/campo.js";
import { formatCpf } from "../lib/cpf.js";
import { calcularIdade } from "../lib/idade.js";
import { rotuloPeriodoInscricao } from "../lib/periodo-inscricao.js";
import { prisma } from "../lib/prisma.js";
import { rotuloSituacaoTurma } from "../lib/situacao-turma.js";
import type { RelatorioAlunosTurmaQuery } from "../validators/relatorio-alunos-turma.js";

const inscricaoInclude = {
  aluno: {
    select: {
      nome: true,
      cpf: true,
      dtNascimento: true,
      escolaridade: { select: { descricao: true } },
    },
  },
  _count: { select: { atendimentos: true } },
} satisfies Prisma.InscricaoAlunoCursoInclude;

type InscricaoRow = Prisma.InscricaoAlunoCursoGetPayload<{
  include: typeof inscricaoInclude;
}>;

type SituacaoAlunoCodigo =
  | "MATRICULADO"
  | "MATRICULA_CANCELADA"
  | "A_MATRICULAR";

function codigoSituacaoAluno(row: {
  matriculado: boolean;
  matriculaCancelada: boolean;
}): SituacaoAlunoCodigo {
  if (row.matriculaCancelada) return "MATRICULA_CANCELADA";
  if (row.matriculado) return "MATRICULADO";
  return "A_MATRICULAR";
}

function rotuloSituacaoAluno(codigo: SituacaoAlunoCodigo): string {
  switch (codigo) {
    case "MATRICULADO":
      return "Matriculado";
    case "MATRICULA_CANCELADA":
      return "Matrícula cancelada";
    case "A_MATRICULAR":
      return "A matricular";
  }
}

function dataMatricula(row: InscricaoRow): string | null {
  if (row.dtInicioCurso != null) return formatDataBr(row.dtInicioCurso);
  if (row.dataHoraMatricula != null) return formatDataBr(row.dataHoraMatricula);
  return null;
}

function dataCancelamento(row: InscricaoRow): string | null {
  if (row.dataHoraCancelamentoMatricula == null) return null;
  return formatDataBr(row.dataHoraCancelamentoMatricula);
}

function dataUltSituacao(row: InscricaoRow): string {
  const codigo = codigoSituacaoAluno(row);
  if (codigo === "MATRICULA_CANCELADA") {
    return dataCancelamento(row) ?? formatDataBr(row.dtCurso);
  }
  if (codigo === "MATRICULADO") {
    return dataMatricula(row) ?? formatDataBr(row.dtCurso);
  }
  return formatDataBr(row.dtCurso);
}

function mapAluno(row: InscricaoRow) {
  const situacaoCodigo = codigoSituacaoAluno(row);
  return {
    inscricaoId: row.id,
    alunoNome: row.aluno.nome,
    alunoCpf: row.aluno.cpf,
    alunoCpfFormatado: row.aluno.cpf ? formatCpf(row.aluno.cpf) : null,
    alunoIdade:
      row.aluno.dtNascimento != null
        ? calcularIdade(row.aluno.dtNascimento)
        : null,
    escolaridadeDescricao: row.aluno.escolaridade?.descricao ?? null,
    dataInscricao: formatDataBr(row.dtCurso),
    dataMatricula: dataMatricula(row),
    dataCancelamento: dataCancelamento(row),
    situacao: rotuloSituacaoAluno(situacaoCodigo),
    situacaoCodigo,
    dataUltSituacao: dataUltSituacao(row),
    quantidadeAtendimentos: row._count.atendimentos,
  };
}

function contarPorSituacao(rows: InscricaoRow[]) {
  let matriculados = 0;
  let matriculasCanceladas = 0;
  let aMatricular = 0;
  for (const row of rows) {
    const codigo = codigoSituacaoAluno(row);
    if (codigo === "MATRICULADO") matriculados++;
    else if (codigo === "MATRICULA_CANCELADA") matriculasCanceladas++;
    else aMatricular++;
  }
  return { matriculados, matriculasCanceladas, aMatricular };
}

function filtrarPorSituacao(
  rows: InscricaoRow[],
  situacao: RelatorioAlunosTurmaQuery["situacao"],
) {
  if (situacao === "TODAS") return rows;
  return rows.filter((row) => codigoSituacaoAluno(row) === situacao);
}

async function listarTurmas(filtros: RelatorioAlunosTurmaQuery) {
  if (filtros.turmaCodigo != null) {
    const turma = await prisma.turma.findFirst({
      where: {
        codigo: filtros.turmaCodigo,
        ativo: true,
        ...(filtros.cursoCodigo != null
          ? { cursoCodigo: filtros.cursoCodigo }
          : {}),
      },
      include: {
        curso: { select: { codigo: true, descricao: true } },
      },
    });
    return turma ? [turma] : [];
  }

  return prisma.turma.findMany({
    where: {
      ativo: true,
      ...(filtros.cursoCodigo != null
        ? { cursoCodigo: filtros.cursoCodigo }
        : {}),
    },
    include: {
      curso: { select: { codigo: true, descricao: true } },
    },
    orderBy: [
      { curso: { descricao: "asc" } },
      { codigo: "asc" },
    ],
  });
}

export async function gerarRelatorioAlunosTurma(
  filtros: RelatorioAlunosTurmaQuery,
) {
  const turmas = await listarTurmas(filtros);

  if (filtros.turmaCodigo != null && turmas.length === 0) {
    throw new RelatorioAlunosTurmaError("Turma não encontrada para o curso informado");
  }

  const turmaCodigos = turmas.map((t) => t.codigo);
  const inscricoesPorTurma = new Map<number, InscricaoRow[]>();

  if (turmaCodigos.length > 0) {
    const todasInscricoes = await prisma.inscricaoAlunoCurso.findMany({
      where: {
        ativo: true,
        turmaCodigo: { in: turmaCodigos },
      },
      include: inscricaoInclude,
      orderBy: [{ aluno: { nome: "asc" } }, { codigo: "asc" }],
    });

    for (const row of todasInscricoes) {
      const lista = inscricoesPorTurma.get(row.turmaCodigo) ?? [];
      lista.push(row);
      inscricoesPorTurma.set(row.turmaCodigo, lista);
    }
  }

  const secoes = turmas.map((turma) => {
    const todas = inscricoesPorTurma.get(turma.codigo) ?? [];
    const filtradas = filtrarPorSituacao(todas, filtros.situacao);
    const resumo = contarPorSituacao(todas);

    return {
      cursoCodigo: turma.curso.codigo,
      cursoDescricao: turma.curso.descricao,
      turmaCodigo: turma.codigo,
      turmaNome: turma.nome,
      periodo: turma.periodo,
      periodoRotulo: rotuloPeriodoInscricao(turma.periodo),
      situacaoTurma: turma.situacao,
      situacaoTurmaRotulo: rotuloSituacaoTurma(turma.situacao),
      alunos: filtradas.map(mapAluno),
      resumo: {
        matriculados: resumo.matriculados,
        matriculasCanceladas: resumo.matriculasCanceladas,
        ...(turma.situacao === "ABERTA"
          ? { aMatricular: resumo.aMatricular }
          : {}),
      },
    };
  });

  return {
    filtros: {
      cursoCodigo: filtros.cursoCodigo ?? null,
      turmaCodigo: filtros.turmaCodigo ?? null,
      situacao: filtros.situacao,
      todosCursos: filtros.cursoCodigo == null,
      todasTurmas: filtros.turmaCodigo == null,
    },
    secoes,
    resumoGeral: secoes.map((s) => ({
      cursoCodigo: s.cursoCodigo,
      cursoDescricao: s.cursoDescricao,
      turmaCodigo: s.turmaCodigo,
      turmaNome: s.turmaNome,
      periodoRotulo: s.periodoRotulo,
      situacaoTurma: s.situacaoTurma,
      situacaoTurmaRotulo: s.situacaoTurmaRotulo,
      matriculados: s.resumo.matriculados,
      matriculasCanceladas: s.resumo.matriculasCanceladas,
      ...(s.situacaoTurma === "ABERTA"
        ? { aMatricular: s.resumo.aMatricular ?? 0 }
        : {}),
    })),
  };
}

export class RelatorioAlunosTurmaError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RelatorioAlunosTurmaError";
  }
}
